### DISCUSSION TITLE
How to put a responsive-height vega visual into a dashboard

### DISCUSSION THREAD

**Discussion by declann** [quarto-dev/quarto-cli#8343] [Q&A] [answered]
### Description

Hi,

I'm making a dashboard including vega visuals. Given the nice responsiveness of dashboard layouts, it will be cool if Vega can use full space that layout calculates for a card.

I'm trying this here:
https://github.com/declann/quarto-dashboard-layout-experiments/blob/main/dash-2.qmd

I have responsive width working with a little work: I specify `width:"container"` in vega spec and via CSS give the `vega-embed` class 100% width, also 100% height. I also trigger a resize event when the vega visual is ready, something noted in vega docs [here](https://vega.github.io/vega-lite/docs/size.html#specifying-responsive-width-and-height).

**This gives me a chart that responds to fill the full width of the card and it works nicely, but height is still fixed to something calculated by vega.**

Adding `height: "container"` should indicate to vega to use the full container height (if available) - as with width, but when I use that here is the result:

![image](https://github.com/quarto-dev/quarto-cli/assets/353214/7f220c8f-3792-40ae-ab98-c0bee7d4e9a3)

Something far stranger is happening because whenever I resize the window, I get to see the bar chart because it begins to get more space for every resize event, but this doesn't seem to make any sense!

If `height: 100%` doesn't work, any ideas to set container height to match card and get a responsive height?

I'll get to do more digging into this in due course, but if someone already understands whats going on and has suggestions I'd love to hear about it!

Thanks
Declan

**Comment by mcanouil** [comment]
(This is unrelated to Quarto.)
For something to occupy 100% it needs to have a dimension first.
If the outer container has no height, 100% of nothing means nothing.
See <https://developer.mozilla.org/en-US/docs/Web/CSS/height> and other CSS related resources online.

In your case, the issue is the "canvas" object you generate which has an absolute height set.
You can easily see that using your browser developer mode.
<img width="1624" alt="Screenshot 2024-01-18 at 18 11 07" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/fb97d5d4-7fc1-4fa4-af21-5921bb649804">

You need to look into vega "height" attribute what values are allowed and set the right one.
See <https://vega.github.io/vega/docs/specification/>.
For support regarding this, see <https://github.com/vega/vega> and/or <https://github.com/vega/vega-embed>.

**Reply by declann** [reply, marked as answer]
Thanks Mikael, it's the mechanism to get the height set that sent me down a rabbithole and I have a tonne to learn about CSS because there has to be a far more elegant way, but

I got a solution here:

https://github.com/declann/quarto-dashboard-layout-experiments/blob/main/dash-3.qmd

by setting numerous parent heights to 100%:

```css
svg, .chart-wrapper, .vega-embed, .vega-embed-wrapper, .observablehq, /*.ojs-cell-2-2,*/ .html-fill-item, .cell-output, .card, .bslib-grid {
  height:100%
}
```

after a `100vh` body height:

```css
body {
  height: 100vh;
}
```

It works neatly in that case* but got ridiculous to apply in my actual usecase because additional grid settings kicked in, which I could force off, but mobile also became an issue whenever cards wrap.

So I'm using fixed height again, and glad to have that CSS layout away from my hands once again!

*but not on mobile https://quarto-dashboard-layout-experiments.pages.dev/dash-3
